src/v/datalake/translation_task.cc

```diff
@@ -0,0 +1,182 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/redpanda-data/redpanda/blob/master/licenses/rcl.md
+ */
+
+#include "datalake/translation_task.h"
+
+#include "datalake/logger.h"
+#include "datalake/record_multiplexer.h"
+
+#include <seastar/core/loop.hh>
+#include <seastar/core/seastar.hh>
+namespace datalake {
+namespace {
+remote_path calculate_remote_path(
+  const local_path& local_file_path, const remote_path& remote_path_prefix) {
+    auto f_name = local_file_path().filename();
+    return remote_path{remote_path_prefix() / f_name};
+}
+
+translation_task::errc map_error_code(cloud_data_io::errc errc) {
+    switch (errc) {
+    case cloud_data_io::errc::file_io_error:
+        return translation_task::errc::file_io_error;
+    case cloud_data_io::errc::cloud_op_error:
+    case cloud_data_io::errc::cloud_op_timeout:
+        return translation_task::errc::cloud_io_error;
+    }
+}
+
+} // namespace
+translation_task::translation_task(cloud_data_io& cloud_io)
+  : _cloud_io(&cloud_io) {}
+
+ss::future<
+  checked<coordinator::translated_offset_range, translation_task::errc>>
+translation_task::translate(
+
+  std::unique_ptr<data_writer_factory> writer_factory,
+  model::record_batch_reader reader,
+  const remote_path& remote_path_prefix,
+  retry_chain_node& rcn,
+  lazy_abort_source lazy_as) {
+    record_multiplexer mux(std::move(writer_factory));
+    // Write local files
+    auto mux_result = co_await reader.consume(
+      std::move(mux), _read_timeout + model::timeout_clock::now());
+
+    if (mux_result.has_error()) {
+        vlog(
+          datalake_log.warn,
+          "Error writing data files - {}",
+          mux_result.error());
+        co_return errc::file_io_error;
+    }
+    auto write_result = std::move(mux_result.value());
+    vlog(
+      datalake_log.trace,
+      "translation result base offset: {}, last offset: {}, data files: {}",
+      write_result.start_offset,
+      write_result.last_offset,
+      fmt::join(write_result.data_files, ", "));
+
+    coordinator::translated_offset_range ret{
+      .start_offset = mux_result.value().start_offset,
+      .last_offset = mux_result.value().last_offset,
+    };
+    ret.files.reserve(write_result.data_files.size());
+    std::optional<errc> upload_error;
+    // TODO: parallelize uploads
+    for (auto& lf_meta : write_result.data_files) {
+        auto r = co_await execute_single_upload(
```

Comment:
question for my understanding, are there any knobs for rate limiting datalake uploads? (so it doesn't interfere with TS uploads).

Reply:
Not really. There are some knobs in tiered storage to slow down housekeeping work, but no knobs for throttling uploads specifically (minus backoff/retries from retry chain node)